libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp

```diff
@@ -215,7 +215,19 @@ bool NavEKF2_core::resetHeightDatum(void)
     stateStruct.position.z = 0.0f;
     // adjust the height of the EKF origin so that the origin plus baro height before and after the reset is the same
     if (validOrigin) {
-        ekfGpsRefHgt += (double)oldHgt;
+        const AP_GPS &gps = AP::gps();
+        if (gps.status() >= AP_GPS::GPS_OK_FIX_3D) {
+            // if we have a good GPS lock then reset to the GPS
+            // altitude. This ensures the reported AMSL alt from
+            // getLLH() is equal to GPS altitude, while also ensuring
+            // that the relative alt is zero
+            ekfGpsRefHgt = gps.location().alt*0.01;
+        } else {
+            // if we don't have GPS lock then we shouldn't be doing a
+            // resetHeightDatum, but if we do then the best option is
+            // to maintain the old error
+            ekfGpsRefHgt += oldHgt;
+        }
```

Comment:
This is the code I would propose for this:

```
    if (inFlight || !gpsGoodToAlign) {
        EKF_origin.alt += (int32_t)(100.0f * oldHgt);
    } else {
        EKF_origin.alt = AP::gps().location().alt*0.01;
   }
    ekfGpsRefHgt = (double)0.01 * (double)EKF_origin.alt;
```

The reason for setting EKF_origin.alt is that it matters when EK2_OGN_HGT_MASK bit 2 is set to 1 (enabling offset to be applied to local position).

Reply:
There are some other changes that are required before this can be done, eg:

- Move the setting of gpsGoodToAlign to inside the calcGpsGoodToAlign() function
- modify readGpsData() so that calcGpsGoodToAlign() is always called.
- modify calcGpsGoodToAlign() so that gpsGoodToAlign is set to true at the top of the function if in air and ekf is operating in its absolute position mode (is aligned). This change needs to preserve the exisiting logic that handles the case of planes that are launched without magnetometer.

Reply:
@priseborough I have made the changes. I added a timeout for switching from good GPS to bad as well as the scaling factor change. I think this is needed to prevent us annoying users with short GPS glitches.
@OXINARF I did try your approach to setting via EKF_original.alt but it didn't work at all. The altitude goes completely crazy

Reply:
@tridge Do you have the log of that? This isn't worse than what we have, but if the local position bit is set, I'm pretty sure the output will be bugged.

Reply:
@OXINARF it happens immediately in SITL with default params. Alt goes between 0 and 650m rapidly

Reply:
@tridge There is an error in my code above, the else branch shouldn't have the multiplication. It should be (comments omitted):

```
    if (!gpsGoodToAlign) {
        EKF_origin.alt += (int32_t)(100.0f * oldHgt);
    } else {
        EKF_origin.alt = AP::gps().location().alt;
   }
    ekfGpsRefHgt = (double)0.01 * (double)EKF_origin.alt;
```

I've tried this in SITL and it looked fine. However, I'm not very experienced with testing Plane in SITL so I might be doing something wrong. Can you please try that code?

One other question is if we should be getting the altitude directly from GPS or using the last altitude EKF got from GPS (data from GPS library can be slightly more recent and not be good).

I've caught another bug in the corrected local position mode though:

https://github.com/ArduPilot/ardupilot/blob/104164d2e44e4d9ba3851d74e15338c5596c8223/libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp#L548

Should be:

```
if ((frontend->_originHgtMode & (1<<2)) == 0) {
  gpsDataNew.hgt = (float)((double)0.01 * (double)gpsloc.alt - ekfGpsRefHgt);
} else {
  gpsDataNew.hgt = 0.01 * (gpsloc.alt - EKF_origin.alt);
}
```